source/distributed/error_predictor.cc

```diff
@@ -0,0 +1,363 @@
+                  (gamma_h * std::pow(.5, future_fe_degree));
+
+                // If the future fe index differs from the active one, also take
+                // into account p-adaptation.
+                if (cell->future_fe_index_set())
+                  *predicted_error_it *=
+                    std::pow(gamma_p,
+                             static_cast<int>(future_fe_degree -
+                                              cell->get_fe().degree));
+
+                break;
+              }
+
+            case parallel::distributed::Triangulation<dim,
+                                                      spacedim>::CELL_COARSEN:
+              {
+                // First figure out which finite element will be assigned to the
+                // parent cell after h-adaptation analogously to
+                // dealii::internal::hp::DoFHandlerImplementation::Implementation::
+                //   collect_fe_indices_on_cells_to_be_refined()
+                std::set<unsigned int> fe_indices_children;
+                for (unsigned int child_index = 0;
+                     child_index < cell->n_children();
+                     ++child_index)
+                  {
+                    const auto child = cell->child(child_index);
+                    Assert(child->is_active() && child->coarsen_flag_set(),
+                           typename dealii::Triangulation<
+                             dim>::ExcInconsistentCoarseningFlags());
+
+                    fe_indices_children.insert(child->future_fe_index());
+                  }
+
+                const unsigned int future_fe_index =
+                  dof_handler->get_fe_collection().find_dominated_fe_extended(
+                    fe_indices_children, /*codim=*/0);
+
+                const unsigned int future_fe_degree =
+                  dof_handler->get_fe_collection()[future_fe_index].degree;
+
+                // Then determine the actually contirbution to the predicted
+                // error of every single cells that is about to be coarsened.
+                float sqrsum_of_predicted_errors = 0.;
+                float predicted_error            = 0.;
+                int   degree_difference          = 0;
+                for (unsigned int child_index = 0;
+                     child_index < cell->n_children();
+                     ++child_index)
+                  {
+                    const auto child = cell->child(child_index);
+
+                    predicted_error =
+                      (**estimated_error_it)[child->active_cell_index()] /
+                      (gamma_h * std::pow(.5, future_fe_degree));
+
+                    degree_difference =
+                      future_fe_degree - child->get_fe().degree;
+
+                    if (degree_difference != 0)
+                      predicted_error *= std::pow(gamma_p, degree_difference);
+
+                    sqrsum_of_predicted_errors +=
+                      predicted_error * predicted_error;
```

Comment:
This and all of the similar formulas should be documented somewhere. Is this in the non-`p::d` class? Could we refer to these formulas in the place where the various gamma factors are passed in as argument? Or say something like "The exact details of how the error is predicted are described in the documentation of class X::Y."

Reply:
https://github.com/dealii/dealii/blob/3bc6103128d0cf3ad01023273f827bd9615e2e8c/include/deal.II/distributed/error_predictor.h#L62-L63

I hoped that these lines would suffice. I'll be more verbose on that.